### PR TITLE
doc/design: add proposal for one-shot sinks

### PR DESCRIPTION
**Pull request by aljoscha**
[View rendered](https://github.com/MaterializeInc/materialize/blob/a8312a19d2ab1e6120c289284e0f0b72b335920a/doc/developer/design/20210608_one_shot_sinks.md)

Some of our users might want and easy way of getting the current data in
a view out of Materialize, without worrying about a continuous stream of
changes, exactly once, and changing topic names or file names. This is
at odds with how current sinks work and at odds with potential stricter
requirements of what I'm retroactively calling _continuous sinks_.

I propose to add the concept of _one-off sinks_, which only write out
data up to a point and then finish. They should have relaxed
requirements and should be easier to implement in the future.

### CODE REVIEW

File: doc/developer/design/20210608_one_shot_sinks.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+## Summary
+
+Some of our users might want and easy way of getting the current data in a view
+out of Materialize, without worrying about a continuous stream of changes,
+exactly once, and changing topic names or file names. This is at odds with how
+current sinks work and at odds with potential stricter requirements of what I'm
+retroactively calling _continuous sinks_.
+
+I propose to add the concept of _one-off sinks_, which only write out data up
```
Comment (uce) [inline doc/developer/design/20210608_one_shot_sinks.md:11]:
```suggestion
I propose to add the concept of _one-shot sinks_, which only write out data up
```

File: doc/developer/design/20210608_one_shot_sinks.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
+# Ephemeral, one-shot sinks for Materialize
+
+## Summary
+
+Some of our users might want and easy way of getting the current data in a view
```
Comment (umanwizard) [inline doc/developer/design/20210608_one_shot_sinks.md:5]:
```suggestion
Some of our users might want an easy way of getting the current data in a view
```

File: doc/developer/design/20210608_one_shot_sinks.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+## Non-Goals
+
+- Change the behavior of existing sinks.
+- Add any new sinks, formats, envelopes, what have you.
+
+## Description
+
+Currently, Materialize knows only one type of sink: _continuous sinks_. They
+continuously write new data when the sinked relation changes. When Materialize
+is restarted continuous sinks are restarted as well and they continue producing
```
Comment (umanwizard) [inline doc/developer/design/20210608_one_shot_sinks.md:31]:
```suggestion
is restarted, continuous sinks are restarted as well, and they continue producing
```

**Review by umanwizard — COMMENTED** [review]
minor prose style nits

**Comment by danhhz** [comment]
A one-shot thing that's explicitly ignored on restart doesn't really seem like a SINK to me, it seems more like a SELECT with a different output than pgwire. crdb solved this same problem by adding an EXPORT which was like SELECT but let you specify some target that it wrote to, which might be a better model for what you're trying to do here? https://www.cockroachlabs.com/docs/v21.1/export.html

OOC, what sinks exist besides kafka? I quickly scanned the docs and saw Avro OCF, any others?

**Comment by umanwizard** [comment]
@danhhz the set of sinks we support is Avro/Kafka, Avro/OCF, and `TAIL` (which SQL folks might claim is not a sink -- but from my perspective, it looks like a sink internally)

**Comment by aljoscha** [comment]
I updated this based on out-of-band discussions.

**Review by uce — COMMENTED** [review]
I really like that we want to be more explicit about our sink properties. 👍🏼 


* I like the name `one-shot`, because it captures the fact that each `COPY TO` requires a separate sink instantiation. On the other hand, I'm wondering whether the relaxed consistency guarantees are the more interesting bit about `one-shot` sinks (if so we might want to reflect that in how we call things).
* Can we imagine a use case for `one-shot` sinks that provide stricter guarantees, e.g. I want a consistent one-time snapshot of my view to S3? If so, we might want to be explicit with the distinction between `one-shot vs. continuous` and `volatile vs. non-volatile` (for lack of a better term).
* We should explicitly point out breakage of backwards compatibility if we limit some existing sinks to `COPY TO` in the future.

File: doc/developer/design/20210608_one_shot_sinks.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+Where `<sink definition>` should in theory be compatible with both `CREATE
+SINK` and `COPY TO` but the sink configuration will determine if a sink is
+usable as a continuous sink or one-shot sink. This is similar to volatile
+sources, where the volatility label is not necessarily attached to a specific
+sink type but depends on the concrete instance configuration.
+
+### Unified Sink Pipeline
+
+The actual implementation of sinks will not differ between the sink types. In
+the long run, we should probably move all sink like concepts to be considered
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply doc/developer/design/20210608_one_shot_sinks.md:85]: Current `SELECT ...` (which you could see as a sink) and `COPY TO STDOUT` are not sinks internally.

File: doc/developer/design/20210608_one_shot_sinks.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+The new proposed syntax for `COPY TO` is:
+
+```sql
+COPY (query) TO <sink definition>
+```
+
+Where `<sink definition>` should in theory be compatible with both `CREATE
+SINK` and `COPY TO` but the sink configuration will determine if a sink is
+usable as a continuous sink or one-shot sink. This is similar to volatile
```
[the remark this answers is not in the document]
Comment (aljoscha) [reply doc/developer/design/20210608_one_shot_sinks.md:78]: It's up to us, but I think `continuous` sinks can easily be `one-shot`, while the other way round is not as easy.

And yes, it does break backwards compatibility of the Sink DDL. (Is it DDL? 🤔)

**Comment by aljoscha** [comment]
Added another section that clearly spells out the immediate changes / breakage:

```
### Immediate User-facing Changes / Breaking Changes

- Avro OCF sinks are not usable with `CREATE SINK` anymore, only write data up
  to "now" and then finish
- Kafka sinks without exactly-once/consistency are not usable with `CREATE SINK
  anymore, only write data up to "now" and then finish
- We need to massage the docs to treat `COPY TO` and `CREATE SINK` together but
  not confuse people too much
```

**Comment by awang** [comment]
> OOC, what sinks exist besides kafka? I quickly scanned the docs and saw Avro OCF, any others?

The 3 "sinks" we have today are: (1) Kafka (2) File (Kafka OCF) (3) TAIL

**Comment by umanwizard** [comment]
It does still seem useful to allow people to specify an egress operation that runs for as long as the system is up (as TAIL does) but for which we provide no guarantees about exactly-onceness, restarts, etc.

If someone has their own way of dealing with these issues, we should still allow them to, for example, continuously dump to an OCF file as we do now with avro OCF sinks. I think it would be unfortunate if we got rid of this ability and only allowed dumps until "now" .

**Comment by aljoscha** [comment]
> It does still seem useful to allow people to specify an egress operation that runs for as long as the system is up (as TAIL does) but for which we provide no guarantees about exactly-onceness, restarts, etc.

I think that makes sense, yes! Then this proposal would become:

 - you can create sinks with either `CREATE SINK` or `COPY TO`
 - creating a sink with `CREATE SINK` will create a sink that persists across restarts, this is it will continuously keep running
 - creating a sink with `COPY TO` will create a finite, one-shot sink that will write up to "now" and then finish
 - some sinks will not behave "well" when used with `CREATE SINK`. Think exactly-once, restarts, correctness, what have you

Do you want to reserve the "right" to create some new sinks that only work with `COPY TO`, because that's easier to implement and comes with less (sometimes implied) requirements?

File: doc/developer/design/20210608_one_shot_sinks.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (uce) [reply doc/developer/design/20210608_one_shot_sinks.md:78]: Thanks for the updated backwards compatibility notes.

**Review by uce — APPROVED** [review]

**Comment by aljoscha** [comment]
I'm merging this now, we can shake out the specifics of naming and documentation in the implementation PR(s). The feedback seemed positive in other channels as well. Thanks!

**Pull request merged**